nodejs/server.js

```diff
@@ -167,4 +169,116 @@ app.all('/', function (req, res) {
 
 });
 
-app.listen(argv.port);
+let server = require('http').createServer({
+    perMessageDeflate: {
+        zlibDeflateOptions: {
+            // See zlib defaults.
+            chunkSize: 1024,
+            memLevel: 7,
+            level: 3
+        },
+        zlibInflateOptions: {
+            chunkSize: 10 * 1024
+        },
+        // Other options settable:
+        clientNoContextTakeover: true, // Defaults to negotiated value.
+        serverNoContextTakeover: true, // Defaults to negotiated value.
+        serverMaxWindowBits: 10, // Defaults to negotiated value.
+        // Below options specified as default values.
+        concurrencyLimit: 10, // Limits zlib concurrency for perf.
+        threshold: 1024 // Size (in bytes) below which messages
+        // should not be compressed.
+    }
+});
+
+// Also mount the app here
+server.on('request', app);
+
+const websocketEvent = {
+    url: 'http://localhost:8000/websocket'
+  } 
+
+const inactiveEvent = {
+    url: 'http://localhost:8000/inactive'
+  } 
+
+
+// Create web socket server on top of a regular http server
+let wss = new WSServer({
+    server: server
+});
+
+
+function noop() {}
+
+function heartbeat() {
+  this.isAlive = true;
+}
+
+
+wss.on('connection', function connection(ws) {
+    // On successful request, there's no body returned 
+    request(websocketEvent, (err, res)=>{
+        if (err || res.statusCode!=200){
+            if (err){
+            console.log(err)
+            } else{
+            console.log("Unexpected response")
+            }
+            ws.send("Error")
+            return 
+        }
+    })
+    ws.isAlive = true;
+    ws.on('pong',heartbeat);
+
+    const interval = setInterval(function ping() {
+        // We check if all connections are alive
+        wss.clients.forEach(function each(ws) {
+        // We check if all connections are alive
+          if (ws.isAlive === false) return ws.terminate();
+      
+          ws.isAlive = false;
+          // If client replies, we execute the hearbeat function and set the connection as active
+          ws.ping(noop);  
+        });
+        // After we have pinged all clients and verified number of active connections is 0, we generate event for inactivity on the websocket
+        if (wss.clients.size == 0){
+            // On successful request, there's no body returned 
+            request(inactiveEvent, (err, res)=>{
+                if (err || res.statusCode!=200){
+                    if (err){
+                        console.log(err)
+                    } else{
+                        console.log("Unexpected response")
+                    }
+                    ws.send("Error")
+                    return 
+                }
+            })
+        };
+      }, 60000); // TODO: Interval needs to be customized 
```

Comment:
This is more of a style thing but I get a bit nervous about duplicating state that the `ws` library has internally. Is there a way to avoid having to track this manually, and instead just inspecting the state of all the clients each time you need to check if it's time to shutdown? 

```javascript
    const interval = setInterval(function checkForInactivityAndPing() {
        // We check if all connections are alive
        // After we have pinged all clients and verified number of active connections is 0, we generate event for inactivity on the websocket
        if (wss.clients.size > 0 ) {
          for (const socket of wss.client) {
            socket.ping()
          }
        } else {
            // No active clients -- mark this fission pod as inactive
            // On successful request, there's no body returned 
            request(inactiveEvent, (err, res)=>{
                if (err || res.statusCode!=200){
                    if (err){
                        console.log(err)
                    } else{
                        console.log("Unexpected response")
                    }
                }
            })
        };
      }, 60000);
```